hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/request/key/OMKeyCommitRequest.java

```diff
@@ -234,15 +235,15 @@ public OMClientResponse validateAndUpdateCache(OzoneManager ozoneManager,
                 keysToDelete, trxnLogIndex);
       }
 
-      long scmBlockSize = ozoneManager.getScmBlockSize();
-      int factor = omKeyInfo.getReplicationConfig().getRequiredNodes();
+      final long preAllocatedSpace = QuotaUtil.getReplicatedSize(
+          allocatedLocationInfoList.size() * ozoneManager.getScmBlockSize(),
+          omKeyInfo.getReplicationConfig());
       omBucketInfo = getBucketInfo(omMetadataManager, volumeName, bucketName);
       // Block was pre-requested and UsedBytes updated when createKey and
       // AllocatedBlock. The space occupied by the Key shall be based on
       // the actual Key size, and the total Block size applied before should
       // be subtracted.
-      long correctedSpace = omKeyInfo.getReplicatedSize() -
-          allocatedLocationInfoList.size() * scmBlockSize * factor;
+      long correctedSpace = omKeyInfo.getReplicatedSize() - preAllocatedSpace;
```

Comment:
Nice finding @kaijchen 
Looks like we are incrementing used bytes in bucketInfo object. Does that reflected into persisted object?
If so, can we add some test to assert that usedbytes on bucket? It would by great if we can add some coverage around this.

Reply:
Thanks for reviewing @umamaheswararao. I have added a new integration test to cover this.

Reply:
Is there a bug here that this change is fixing, or is this a refactor? I remember making changes around this area recently, and the "pre-allocated" space was just the number of blocks * required nodes * block_size, as we assume the blocks will be filled when we allocated them.

Reply:
This is not a refactor. If we are pre-allocating `blocks * required nodes * block_size` for a EC key, then we have pre-allocated too much space. Either the `preAllocatedSpace` here is incorrect or the `numPreAllocatedBlocks` is incorrect.

Reply:
Can you please explain more where the bug is exactly? `blocks * required nodes * block_size` is the correct pre-allocation, where for rs-3-2, the calculation is `1_block * 5 * 256mb` as that is how much space we reserve assuming the blocks will be filled and including the replication overhead. When the key is committed, we calculate the actual replicated space used and remove the pre-allocated amount.

It might be helpful if you can give an example of what the pre-allocated would be before and after this change and how the calculation is wrong to help me understand this change.

Reply:
What does `block` mean here for EC keys? If it's a `stripe (block group)` then you are correct.
I thought it was individual `block (BlockOutputStreamEntry)` in each container.

Reply:
> where for rs-3-2, the calculation is `1_block * 5 * 256mb` as that is how much space we reserve assuming the blocks will be filled and including the replication overhead.

OK, I got it. The minimum allocation unit in each container is `1_block`, which is `256MB` by default.
The patch above is wrong. But there are still issues in the original code.

Assume the config is `rs-3-2-1024k` and `blockSize = 256 MB`.
| KeySize | PreAllocateSpace (Expected) |  PreAllocateSpace (Current) |
| --- | --- | --- |
| 1 MB | 256 MB * 5 | 256 MB * 5 |
| 257 MB | 256 MB * 5 | 256 MB * 5 * 2 |
| 513 MB | 256 MB * 5 | 256 MB * 5 * 3 |
| 769 MB | 256 MB * 5 * 2 | 256 MB * 5 * 4 |

Reply:
For EC we are allocating in "ec block" chunks, which is always 256 * 5 (rs-3-2). We are not allocating in 256MB chunks. The code `allocatedLocationInfoList.size()` should be returning the number of EC Block groups allocated. Assuming it does, the calculation of pre-allocated size of `allocatedLocationInfoList.size() * blockSize * 5` is correct I think?

In the table above, if the keysize is blockSize (256MB) then we should only need to allocate a single EC block group - the client should never request multiple blocks.

Assuming rs-3-2. The client requests an EC block, which on OM will pre-allocate 256 * 5. The client fills this block with 256 * 3 MB of data + 2 * 256MB of parity. Then it requests another block. This block adds 256 * 5 to the pre-allocated amount. In your table above, the current code should pre-allocate only blockSize * 5 for all rows except the last, where it will allocate `blockSize * 5 * 2`.

The change made, I don't think changes the original logic:

```
// New code
final long preAllocatedSpace = QuotaUtil.getReplicatedSize(
          allocatedLocationInfoList.size() * ozoneManager.getScmBlockSize(),
          omKeyInfo.getReplicationConfig());
      omBucketInfo = getBucketInfo(omMetadataManager, volumeName, bucketName);
long correctedSpace = omKeyInfo.getReplicatedSize() - preAllocatedSpace;

// Old Code
      long correctedSpace = omKeyInfo.getReplicatedSize() -
          allocatedLocationInfoList.size() * scmBlockSize * factor;
```

Inside quota Util, it will take the passed datasize and effectively multiple by factor to get the replicated size, so it looks to me like the logic remains the same?

Reply:
@sodonnel Thanks for the explanation. But seems it's not the case.

> In your table above, the current code should pre-allocate only blockSize * 5 for all rows except the last, where it will allocate blockSize * 5 * 2.

`blockSize = 4194304` and `keySize = 3 * BlockSize + 1` in the screenshot below.
<img width="340" alt="debug screenshot" src="https://user-images.githubusercontent.com/5821159/158405390-5312c797-8de4-4904-b930-9bbff554b531.png">

And can you explain why the last stripe needs `blockSize * 5 * 2`?

> Inside quota Util, it will take the passed datasize and effectively multiple by factor to get the replicated size, so it looks to me like the logic remains the same?

It's not the same, it will treat ECReplicationConfig differently. But the "new code" now is not correct anyway.

Reply:
> It's not the same, it will treat ECReplicationConfig differently. But the "new code" now is not correct anyway.

You are correct. If you pass in 256MB as the data size into QuotaUtil it will stripe it across the blocks and give you the real replicated size. But when we allocate an EC block, pre-allocating blockSize * 5 (for RS-3-2) is the correct thing to do, as that is how much space we might use. It seems there is something not correct in the locationList or how we are using it.